internal/refactoring/cross_provider_move.go

```diff
@@ -0,0 +1,184 @@
+package refactoring
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs/configschema"
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/states"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// crossTypeMover is a collection of data that is needed to calculate the
+// cross-provider move state changes.
+type crossTypeMover struct {
+	State         *states.State
+	ProviderCache map[addrs.Provider]providers.Factory
+}
+
+func (m *crossTypeMover) getProvider(providers addrs.Provider) (providers.Interface, error) {
+	if provider, ok := m.ProviderCache[providers]; ok {
```

Comment:
I don't think this is actually caching anything, since you call the factory function here, and `Close()` the provider after every move. Probably not a big deal for small single moves, but large refactoring with heavy providers could be slow.

Reply:
Done!